pacemaker_stonith: new module

munchtoast opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
* Adds in `pacemaker_stonith` ansible module for managing cluster stonith

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

<!--- Please do not forget to include a changelog fragment:
      https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-changelog-fragments
      No need to include one for docs-only or test-only PR, and for new plugin/module PRs.
      Read about more details in CONTRIBUTING.md.
      -->

##### ISSUE TYPE
<!--- Pick one or more below and delete the rest.
      'Test Pull Request' is for PRs that add/extend tests without code changes. -->
- New Module/Plugin Pull Request

##### COMPONENT NAME
<!--- Write the SHORT NAME of the module, plugin, task or feature below. -->
`pacemaker_stonith`

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

munchtoast commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,218 @@
+        self.runner = pacemaker_runner(self.module, cli_action='stonith')
+        self.vars.set('previous_value', self._get())
+        self.vars.set('value', self.vars.previous_value, change=True, diff=True)
+        self.vars.set('resource_type', self.vars.stonith_type)
+        self.vars.set('resource_option', self.vars.stonith_option)
+        self.vars.set('resource_operation', self.vars.stonith_operation)
+        self.vars.set('resource_meta', self.vars.stonith_meta)
+        self.vars.set('resource_argument', self.vars.stonith_argument)
```
Leaving this PR as a draft as I would like to receive some advice/suggestions on how to approach this. The `module_utils/pacemaker.py` have these variables (i.e. `resource_type`) and I would like to assign it to be the same as `stonith_type` from this module. Unfortunately, that doesn't seem to be possible, so I am wondering if there's another way how I could handle this? Aliases in the module arguments?

munchtoast commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,213 @@
+                'state name resource_type resource_option resource_operation resource_meta resource_argument wait',
+                output_process=self._process_command_output(True, "already exists"),
+                check_mode_skip=True) as ctx:
+            ctx.run(resource_type={'resource_type': self.vars.stonith_type}, resource_option={'resource_option': self.vars.stonith_option}, resource_operation={'resource_operation': self.vars.stonith_operation}, resource_meta={'resource_meta': self.vars.stonith_meta}, resource_argument={'resource_argument': self.vars.stonith_argument})
```
This gets the module to "work" but this is not a clean way to handle this..

munchtoast commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,218 @@
+        self.module.params['resource_option'] = self.vars.stonith_option
+        self.module.params['resource_operation'] = self.vars.stonith_operation
+        self.module.params['resource_meta'] = self.vars.stonith_meta
+        self.module.params['resource_argument'] = self.vars.stonith_argument
```
The original way I had was passing in the arguments directly into `ctx.run()` which I'm not sure if that was the right option as well

russoz commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,218 @@
+  - Dexter Le (@munchtoast)
+version_added: 10.8.0
+description:
+  - This module can manage stonith in a Pacemaker cluster using the pacemaker CLI.
```
Not familiar with "stonith" but it looks like a name of a product/service/something. Shouldn't it be capitalized throughout the docs?

felixfontein replied:
If it's about the technique (https://en.wikipedia.org/wiki/STONITH#STONITH), it should be written "STONITH".

munchtoast commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,218 @@
+        description:
+          - Operation action to associate with stonith.
+        type: str
+      operation_option:
+        description:
+          - Operation option to associate with action.
+        type: list
```
Good catch.. This will also need to be updated in a future PR: https://github.com/ansible-collections/community.general/blob/main/plugins/modules/pacemaker_resource.py#L72

russoz commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,228 @@
+        supports_check_mode=True
+    )
+
+    use_old_vardict = False
```
No longer needed from c.g. 11.0.0 onwards ;-)

munchtoast replied:
```suggestion

```

russoz commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,228 @@
+    )
+
+    use_old_vardict = False
+    default_state = "present"
```
Since `state` will always have some value, and its default is `present`, this is kinda redundant.

For the record - I just noticed this and have made that same mistake in many of my own modules ;-)

russoz commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,228 @@
+        with self.runner('state name', output_process=self._process_command_output(False)) as ctx:
+            return ctx.run(state='status')
+
+    def _fmt_stonith_resource(self):
```
This file is not meant to be imported by any other. Prefixing the method names with `_` is kinda redundant in this context. In the `module_utils` file that would be valid if not necessary.

russoz commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,228 @@
+                    resource_meta=self.vars.stonith_metas,
+                    resource_argument=self.vars.stonith_argument)
+            self.vars.set('value', self._get())
+            self.vars.stdout = ctx.results_out
+            self.vars.stderr = ctx.results_err
+            self.vars.cmd = ctx.cmd
```
If you do this in the output_process function, no need to repeat yourself - keep it DRY

russoz commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,228 @@
+    def state_absent(self):
+        with self.runner('state name', output_process=self._process_command_output(True, "does not exist"), check_mode_skip=True) as ctx:
+            ctx.run()
+            self.vars.set('value', self._get())
```
You may use `__quit_module__()`, the counterpart of `__init_module__()` and do it only once.

russoz reviewed (commented):
hi @munchtoast 

Couple of comments more :-)

felixfontein commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,228 @@
+        type: str
+      operation_options:
+        description:
+          - Operation option to associate with action.
```
```suggestion
          - Operation options to associate with action.
```

felixfontein commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,228 @@
+        elements: str
+  stonith_metas:
+    description:
+      - List of meta to associate with STONITH.
```
Should this be "metadata"?
```suggestion
      - List of metadata to associate with STONITH.
```

felixfontein commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,228 @@
+        elements: str
+  agent_validation:
+    description:
+      - enabled agent validation for STONITH creation.
```
```suggestion
      - Enabled agent validation for STONITH creation.
```

felixfontein commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,228 @@
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
```
```suggestion
```

felixfontein commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,228 @@
+'''
+
+EXAMPLES = '''
+---
```
```suggestion
```

felixfontein commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,228 @@
+
+EXAMPLES = '''
+---
+- name: Create pacemaker STONITH
+  hosts: localhost
+  gather_facts: false
+  tasks:
+  - name: Create virtual-ip STONITH
+    community.general.pacemaker_stonith:
+      state: present
+      name: virtual-stonith
+      stonith_type: fence_virt
+      stonith_options:
+        - "pcmk_host_list=f1"
+      stonith_operations:
+        - operation_action: monitor
+          operation_options:
+            - "interval=30s"
```
Usually module examples are tasks, not playbooks. Since this playbook has a single task and the playbook part is pretty much boilderplate and unrelated to this example, how about removing it?
```suggestion
- name: Create virtual-ip STONITH
  community.general.pacemaker_stonith:
    state: present
    name: virtual-stonith
    stonith_type: fence_virt
    stonith_options:
      - "pcmk_host_list=f1"
    stonith_operations:
      - operation_action: monitor
        operation_options:
          - "interval=30s"
```

munchtoast commented:
@russoz , @felixfontein 
Sorry for delayed response on my end; I think I've addressed most of the suggestions here. There are some changes in `module_utils/pacemaker.py` noted in: https://github.com/ansible-collections/community.general/pull/10227 that will need to be added here for `cli_action` (Mostly using one runner instead of multiple).

Due to this, I'm going to move this into `Draft` status as I'll need to refactor for that. Thank you for your feedbacks on both PRs!! :)

russoz reviewed (commented):
Minor comments, other than that LGTM.

felixfontein commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,213 @@
+
+DOCUMENTATION = '''
+module: pacemaker_stonith
+short_description: Manage pacemaker STONITH
```
Same here, it should be Pacemaker not pacemaker.
```suggestion
short_description: Manage Pacemaker STONITH
```

felixfontein commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,213 @@
+  - Dexter Le (@munchtoast)
+version_added: 11.2.0
+description:
+  - This module manages STONITH in a Pacemaker cluster using the pacemaker CLI.
```
It's either
```suggestion
  - This module manages STONITH in a Pacemaker cluster using the Pacemaker CLI.
```
or
```suggestion
  - This module manages STONITH in a Pacemaker cluster using the C(pacemaker) CLI.
```
if you're refering to the CLI tool of that name.

russoz commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,213 @@
+                         ('err', result[2])])
+
+    def fmt_stonith_resource(self):
+        return dict([("resource_name", self.vars.stonith_type)])
```
Also

```suggestion
        return dict(resource_name=self.vars.stonith_type)
```
Or

```suggestion
        return {"resource_name": self.vars.stonith_type} 
```
Same on the function below

russoz reviewed (commented):
Spotted a couple of improvements there.

felixfontein commented on plugins/modules/pacemaker_stonith.py:
```diff
@@ -0,0 +1,213 @@
+'''
+
+RETURN = '''
+cluster_stonith:
```
I don't understand where this return value is set in the module.

russoz reviewed (approved):
LGTM
